src/System.Private.Uri/src/System/UriHelper.cs

```diff
@@ -332,232 +363,209 @@ static void EnsureCapacity(char[]? dest, int destSize, int requiredSize)
 
             while (true)
             {
-                // we may need to re-pin dest[]
-                fixed (char* pDest = dest)
+                if ((unescapeMode & UnescapeMode.EscapeUnescape) == UnescapeMode.CopyOnly)
                 {
-                    if ((unescapeMode & UnescapeMode.EscapeUnescape) == UnescapeMode.CopyOnly)
-                    {
-                        while (start < end)
-                            pDest[destPosition++] = pStr[start++];
-                        return dest;
-                    }
+                    while (start < end)
+                        dest.Append(pStr[start++]);
```

Comment:
Seems like this loop would instead be better as:
```C#
dest.Append(new ReadOnlySpan<char>(pStr, end - start));
```
(assuming `end` is >= `start`... if it's not this could be surrounded by an `if` that does that check).

Reply:
I'd prefer not to mix my changes aimed to optimize memory allocations with a refactoring to avoid accidentally breaking the behavior. Specifically, "start" is modified here and accessed in other places in the same large method, all other usages need to be checked before removing post-increment. We can refactor this in a separate PR.

Reply:
> with a refactoring

Ok, but this isn't a refactoring.  You're changing the type into which you're writing data and utilizing the right methods on that type to do so.  Making potentially lots of calls to one overload on the new type when a better overload on the new type exists isn't an orthogonal refactoring to your change.

Reply:
Actually, I tried following your suggestion and converting by-char Append to a span-based operation, but for it to work properly in these two places I need a ctor ReadOnlySpan<char>(char* pointer, **int start**, int length) which doesn't exist. There are only ReadOnlySpan<char>(void* pointer, int length) and ReadOnlySpan<char>(char[]? array, int start, int length).

Reply:
```
dest.Append(new ReadOnlySpan<char>(pStr + start, end - start));
```

Reply:
Or:
```
dest.Append(new ReadOnlySpan<char>(pStr, end - start).Slice(start));
```

Reply:
Oh, pointer arithmetic works in C# the same way as in C. Ok, then, I will try it out.